### ISSUE TITLE
JS backend: bugs involving "null"

### ISSUE THREAD

**Issue by frankymacster** [agda/agda#4962]
I'm trying to use compile Agda to js so that I can use the functions compiled from Agda in [TodoMVC](https://github.com/reduxjs/redux/tree/master/examples/todomvc).

I was able to generate the js code but when I try to call the functions defined [here](https://gist.github.com/frankymacster/0874b7944a4fd043faa9f4fc366656a5) I got this error

<img width="1033" alt="Screen Shot 2020-10-07 at 22 40 14" src="https://user-images.githubusercontent.com/10590239/95408703-173c5280-08ee-11eb-8644-25d137633a87.png">

it then became quite difficult to understand how to "translate" the native javascript data types into Agda compiled data types...

Here is the code I'm dealing with on the js side:

```js
import {
  ADD_TODO,
  DELETE_TODO,
  EDIT_TODO,
  COMPLETE_TODO,
  COMPLETE_ALL_TODOS,
  CLEAR_COMPLETED
} from '../constants/ActionTypes'

import {
  Todo,
  AddTodo,
  DeleteTodo,
  EditTodo,
  CompleteTodo,
  CompleteAllTodos,
  ClearCompleted
} from '../logic/js/jAgda.todos'

import { Bool } from '../logic/js/jAgda.Agda.Builtin.Bool'
import { List } from '../logic/js/jAgda.Agda.Builtin.List'
import { Nat } from '../logic/js/jAgda.Agda.Builtin.Nat'
import { BigInteger } from '../logic/js/biginteger'

const initialState =
  List['_∷_'](Todo.record('Use Redux')(Bool.false)(BigInteger(0)))(List['[]'])

export default function todos(state = initialState, action) {
  switch (action.type) {
    case ADD_TODO:
      return AddTodo(state)(action.text)

    case DELETE_TODO:
      return DeleteTodo(state)(BigInteger(action.id))

    case EDIT_TODO:
      return EditTodo(state)(BigInteger(action.id))(action.text)

    case COMPLETE_TODO:
      return CompleteTodo(state)(BigInteger(action.id))

    case COMPLETE_ALL_TODOS:
      return CompleteAllTodos(state)

    case CLEAR_COMPLETED:
      return ClearCompleted(state)

    default:
      return state
  }
}
```

Here are my questions:
1. I'm wondering if there's a known mapping between the js native data types and Agda compiled-to-js `Agda.Builtin.*` data types.
2. I also noticed that I had some Agda compiled-to-js `Agda.Builtin.*` have some exported fields that are `undefined`
  ```js
  exports["Nat"] = {};
  exports["Nat"]["zero"] = function (x0) {
      return x0["zero"]();
    };
  exports["Nat"]["suc"] = function (x0) {
      return function (x1) {
        return x1["suc"](x0);
      };
    };
  exports["_+_"] = undefined;
  exports["_-_"] = (agdaRTS.primNatMinus);
  exports["_*_"] = undefined;
  exports["_==_"] = undefined;
  ```
  I'm not sure what to do in these cases and could this impact the use of other functions or data types that were Agda compiled-to-js
3. Should I instead look into using the `COMPILED_JS` pragma to define the Agda compiled-to-js data types myself? If so, how do I use `COMPILED_JS`? I'm currently getting these errors in Agda
  ```
  redux/examples/todomvc/src/logic/todos.agda:11,5: Parse error
  COMPILED_JS<ERROR>
  _≟_ function (x) { return fun...
  ```

I'm really glad I found out that Agda has a JS backend for compilation and really hope to use it to interop with js so that the logic can be certified by Agda while the side-effects can be managed in js. I feel I'm very close to having that but the fact that I couldn't find the right mapping between the Agda compiled-to-js data types and the js native data types really makes me hope I made the right choice to go with Agda...

Thanks!

**Comment by frankymacster** [comment]
After further investigation, it turns out that the error I'm seeing is coming from the Agda compiled-to-js in `jAgda.Data.Nat.Properties.js`:
```js
exports["_≟_"] = function (x0) {
    return function (x1) {
      return z_jAgda_Relation_Nullary_Decidable_Core["map′"](agdaRTS.primIntegerFromString("0"))(agdaRTS.primIntegerFromString("0"))(agdaRTS.primIntegerFromString("0"))(agdaRTS.primIntegerFromString("0"))(agdaRTS.primIntegerFromString("0"))(agdaRTS.primIntegerFromString("0"))(z_jAgda_Data_Bool_Properties["T?"]((agdaRTS.uprimIntegerEqual)(x0, x1)));
    };
  };
```

`z_jAgda_Relation_Nullary_Decidable_Core["map′"]` is expecting its 5th argument to be a function but the value coming in is `agdaRTS.primIntegerFromString("0")`.

So the error doesn't seem to be related to which data type I passed in as argument to the compiled from the function I wrote  `DeleteTodo`. Instead it seems related to the compiled files coming from `Data.Nat.Properties` and `Relation.Nullary.Decidable.Core`.

Is there any way to work around this problem?

**Comment by andreasabel** [comment]
Which Agda version are you using @frankymacster >  When I compile `Agda.Data.Nat.Properties` to JS, using Agda-2.6.2 (current master), I see:
```javascript
exports["_≟_"] = a => b => z_jAgda_Relation_Nullary_Decidable_Core["map′"](null)(
    z_jAgda_Data_Bool_Properties["T?"](agdaRTS.uprimIntegerEqual(a,b))
  );
```

**Comment by frankymacster** [comment]
Hi @andreasabel I'm using Agda-2.6.1. Does the JS backend of Agda-2.6.2 actually compile to es6?

**Comment by andreasabel** [comment]
It seems that the JS backend was improved recently, see #3395.  We do not backport fixes to older versions of Agda.

Can you try to install Agda-2.6.2 from the git repo?  This seems to be your best bet.  Or do you need to work with a released version of Agda?

**Comment by frankymacster** [comment]
I'm trying to install Agda from source on my Mac following [these instructions](https://agda.readthedocs.io/en/latest/getting-started/installation.html#installation-of-the-development-version).
When I run `cabal install` I see:
```
Wrote tarball sdist to
/Users/fss/Dropbox/Documents/projects/Coding/dev/agda/agda/dist-newstyle/sdist/Agda-2.6.2.tar.gz
Resolving dependencies...
Build profile: -w ghc-8.10.2 -O1
In order, the following will be built (use -v for more details):
 - Agda-2.6.2 (exe:agda, exe:agda-mode) (requires build)
Starting     Agda-2.6.2 (all, legacy fallback)
Building     Agda-2.6.2 (all, legacy fallback)
```
but it never finishes.
I think I had the same problem when trying to use cabal to install Agda. I ended up using Homebrew and it worked for the released version.
How do I get beyond this point? Is there something specific to do on the Mac for `cabal install` to work properly?

Thank you

**Comment by knrafto** [comment]
I recently had to struggle through this too. Does this work?

```
brew install make
gmake install CABAL_OPTS='-j1 --extra-lib-dirs=/usr/local/opt/icu4c/lib --extra-include-dirs=/usr/local/opt/icu4c/include'
```

Some notes:
* `gmake` from homebrew is necessary because the version of `make` packaged with macos is too old
* I like `-j1` because then cabal will log progress to stdout, which is comforting. but it probably makes the build slower
* You can create the file `mk/config.mk` (it's automatically gitignored) with contents like `CABAL_OPTS='-j1 --extra-lib-dirs=/usr/local/opt/icu4c/lib --extra-include-dirs=/usr/local/opt/icu4c/include'` so you don't have to pass it though the command line each time

I'll try to create a PR to fix the docs.

Edit: after #4980 you can just use the version of `make` that ships with macOS.

**Comment by frankymacster** [comment]
@knrafto I tried the commands above and got:
```
[383 of 383] Compiling Agda.Main        ( src/full/Agda/Main.hs, dist-2.6.2/build/Agda/Main.o, dist-2.6.2/build/Agda/Main.dyn_o )
clang: error: no such file or directory: 'dist-2.6.2/build/Agda/TypeChecking/Generalize.dyn_o'
`gcc' failed in phase `Linker'. (Exit code: 1)
Failed to install Agda-2.6.2
cabal: Error: some packages failed to install:
Agda-2.6.2-BpjyRlQBMorJEW25z7Y2am failed during the building phase. The
exception was:
ExitFailure 1


real    620m8.479s
user    20m40.261s
sys     13m25.320s
gmake: *** [Makefile:129: install-bin] Error 1
```
Anything I can do to fix this? Thanks!

**Comment by andreasabel** [comment]
This might be that the previous compilation was interrupted before this `.dyn_o` file was written.  I'd try to remove the generated files for this module and run compilation again:
```
rm dist-2.6.2/build/Agda/TypeChecking/Generalize.*o
rm dist-2.6.2/build/Agda/TypeChecking/Generalize.hi
gmake install CABAL_OPTS='-j1 --extra-lib-dirs=/usr/local/opt/icu4c/lib --extra-include-dirs=/usr/local/opt/icu4c/include'
```
If this does not work, you can more radically remove the entire build directory and try again.

**Comment by frankymacster** [comment]
Thanks for that. The installation worked! but when I tried to compiling to js I get
`TypeError: Cannot set property '_≈_' of null`

This comes from `./src/logic/js/jAgda.Algebra.Bundles.js`
and is due to
```js
exports["BoundedLattice"] = null;
exports["BoundedLattice"]["_≈_"] = null;
```
I wonder why the compiler does this. Is it how I ran the compiler locally that causes this or does it always compile this way?
Thanks again!

**Comment by frankymacster** [comment]
Anything I can do to get further? Or is this a limitation of the JS backend?

**Comment by andreasabel** [comment]
@frankymacster : I think I fixed the problem with the `= null`, you can try the Agda version at PR #5014 to see whether this solves your problem.
```
cd <your Agda git repo>
git fetch
git checkout js-5002
make install-bin
```

**Comment by frankymacster** [comment]
@andreasabel thanks for looking into this!
I tried with the branch `js-5002`.
I tried creating a list of records in js like so:
```js
[Todo.record('Use Redux')(false)(BigInteger(0))]
```
and now get the error:

```
TypeError: a is not a function

true
src/logic/js/jAgda.Relation.Nullary.Decidable.Core.js:63
  60 |   c,
  61 |   {
  62 |     "false": () => z_jAgda_Relation_Nullary["Reflects"]["ofⁿ"],
> 63 |     "true": () => z_jAgda_Relation_Nullary["Reflects"]["ofʸ"](
     | ^  64 |       a(
  65 |         z_jAgda_Relation_Nullary_Reflects["invert"](d)
  66 | ) ) } )
```

after further debugging I found that 
in `src/logic/js/jAgda.Data.Nat.Properties.js:47`
```
exports["_≟_"] = a => b => z_jAgda_Relation_Nullary_Decidable_Core["map′"](null)(
    z_jAgda_Data_Bool_Properties["T?"](agdaRTS.uprimIntegerEqual(a,b))
  );
```
is being called with `null` as first parameter which seems to be making `z_jAgda_Relation_Nullary_Decidable_Core["map′"]` fail.

I noticed elsewhere that I see some `null`s causing other functions to fail like in
```js
src/logic/js/jAgda.todos.js:47
  44 |       exports["Todo"]["record"](c)(exports["Todo"]["completed"](d))(exports["Todo"]["id"](d))
  45 |     )(d)
  46 |   )(a);
> 47 | exports["CompleteTodo"] = a => b => z_jAgda_Data_List_Base["map"](
  48 |       c => z_jAgda_Data_Bool_Base["if_then_else_"](
  49 |       z_jAgda_Relation_Nullary_Decidable_Core["⌊_⌋"](null)(null)(
  50 |         z_jAgda_Data_Nat_Properties["_≟_"](exports["Todo"]["id"](c))(b)
```

is it normal that these `null`s appear when compiling to js? Thanks again!

**Comment by andreasabel** [comment]
> is it normal that these `null`s appear when compiling to js?

The `null`s are coming from erased things, like type arguments.
However, this does not mean that they are always correctly inserted, there could be a bug.
Most helpful would be a MWE not involving the standard library, i.e. some short Agda code which demonstrates the bug in the JS backend.

> I tried with the branch `js-5002`.

Now merged into master.

**Comment by GregoryMorse** [comment]
I fixed map' by removing the spurious x4 call.  Then I fixed a bug occurring twice where dictionaries were being overwritten with big integers with 0.  The whole stdlib seems to work fine otherwise.

I will share my code as I was able to run client non-node js javascript filling in the undefined functions and using a browser terminal emulator that I designed to simulate IO.  Some beautiful things happened as the async await pattern in Javascript aligns perfectly with the IO primitive _>>=_ and return postulates which drives everything.  Sure if IO does not interact with the user it could run synchronously but at least module loading must be async.  Programs that work in GHC might have infinite loops or not behave correctly though due to strict evaluation.  This is easily fixed with # return Data.Unit.tt >>= \lambda _ -> ... silliness to delay the generating function and its flattening.

Some postulates like primNatToChar or primCharToNat are so easily implemented e.g. string.fromCharCode(num) and string object.charCodeAt(0) that it is hard to understand why implementations of the primitives are not generated.  Even IO could provide it so merely a simple javascript callback interface if provided would be used.

Some other hints are that by dependency sorting all the requires allows simple HTML loading.  By wrapping all functions and objects involving loading or executing any generated code, you can fix the global exports variable and the Proxy object allows dictionaries to be custom modified so the bugs and undefined fillings can be done without touching the generated code at all.  Non node js of course.  Node js has almost certainly alternate solutions for the proxying of exports to patch them.

Not surprisingly after finally getting it all working, it was very little Javascript code to do the entire task.  But it was tricky and time consuming to figure it out.

I want to share it in case it inspires enhancements to the agda javascript generator beyond the bug fixes.  Plus the code is simple and worth a thousand words

**Comment by gallais** [comment]
> Some postulates like primNatToChar or primCharToNat are so easily implemented e.g. string.fromCharCode(num) and string object.charCodeAt(0) that it is hard to understand why implementations of the primitives are not generated. 

Probably because the backend is set up in such a way that we do not get a warning when we
add a new primitive and forget to add the corresponding magic incantation for code generation.

**Comment by jespercockx** [comment]
It seems that the original issue has been fixed by Andreas and it is not clear what issues are still remaining. If you have additional issues with the JS backend, please report them as separate issues.
